bn-apps/cordapp-updates-distribution/design/design.md

```diff
@@ -0,0 +1,135 @@
+![Corda](https://www.corda.net/wp-content/uploads/2016/11/fg005_corda_b.png)
+
+# CordApp distribution service (CDS)
+
+DOCUMENT MANAGEMENT
+---
+
+## Document Control
+
+| Title                | CordApp distribution service                             |
+| -------------------- | ------------------------------------------------------------ |
+| Date                 | 22 Aug 2018                                                |
+| Author               | Ivan Schasny, Mike Hearn |
+| Distribution         | Design Review Board, Product Management, Solutions Engineering, Platform Delivery |
+| Corda target version | OS                                                   |
+
+## HIGH LEVEL DESIGN
+
+### Overview
+
+This proposal describes the architecture of a reference implementation for a CorDapp Distribution Service (CDS), that would allow BNOs to distribute CorDapps updates to Business Network participants. This design targets scenarios when Business Network members maintain their nodes by themselves and can decide on what version of a CorDapp they would like to upgrade to. This design does *not* target scenarios, when upgrade function is centralised and the whole network can be shut down to perform stop-upgrade-restart.
+
+### Requirements
+
+The following requirements have been gathered from various internal discussions within R3 and from [groups.io mailing lists](https://groups.io/g/corda-dev/message/190?p=,,,20,0,0,0::relevance,,updates+distribution,20,2,0,22686107):
+
+* BNO should be able to notify members about new CorDapp version availability
+* BNO should be able to revoke a specific version of CorDapp from their BN
+* A member should be able to download a specific version of CorDapp
+* BNO should be able to collect a report from members about CorDapp versions installed on their nodes
+* Members should be able to subscribe to a repository channel and be aware of it's synchronicity
+* It should be possible to prevent a CorDapp from working if a newer version is available
```

Comment:
what does this mean in practical terms?  That it must be possible to stop somebody's node (which they control) from running a particular piece of code? (How?).  That it must be possible to prevent transactions created by a particular version from being accepted by other nodes? something else?

Reply:
ok - perhaps reword this too?  I think we need to be really clear about when the platform is able to enforce something versus when the platform/CDS code is able to provide the capability for somebody to write an app that enforces something.  ie I think we need to avoid writing things that lead people into incorrect mental models about how computers work :)

Reply:
Rephrased a bit. Hopefully makes more sense now.